[FLINK-16376][yarn] Use consistent method to get Yarn application dir…

jiasheng55 opened the pull request:
…ectory.

## What is the purpose of the change

Currently, the Yarn application directory of Flink is "/user/{user.name}/.flink", but this logic is separated in different places. This PR aims to improve this.

## Brief change log

  - Move `getYarnFilesDir` method from `YarnClusterDescriptor` to `Utils`.
  - Use `getYarnFilesDir` to get target directory of shipped files.

## Verifying this change

This change is already covered by existing tests, such as *org.apache.flink.yarn.YarnFileStageTest*.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (*yes, Yarn deployment*)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

kl0u commented on flink-yarn/src/test/java/org/apache/flink/yarn/YarnFileStageTestS3ITCase.java:
```diff
@@ -161,10 +161,8 @@ private void testRecursiveUploadForYarn(String scheme, String pathSuffix) throws
-
 			YarnFileStageTest.testCopyFromLocalRecursive(fs.getHadoopFileSystem(),
-				new org.apache.hadoop.fs.Path(directory.toUri()), Path.CUR_DIR, tempFolder, false);
+				Path.CUR_DIR, tempFolder, false);
```
@jiasheng55 The `YarnFileStageTestS3ITCase .testRecursiveUploadForYarnS3a ()` seems to be failing on my Travis and the failure seems to be related to this change here. Could you please fix it and then I can merge.

jiasheng55 replied:
@kl0u Sorry I didn't reproduce the failure on my Travis and I don't know how to run `YarnFileStageTestS3ITCase` manually, could you tell me how to reproduce this failure?

kl0u replied:
@jiasheng55 in general we are trying to avoid using `mockito` in our tests, so I suggest to try to get rid of it here.

jiasheng55 commented:
@flinkbot run travis

kl0u commented:
Hi @jiasheng55 . The problem is that to reproduce it you need S3 credentials. I will have another look at the PR but are you sure you do not change anything in the path used by the test? Can you print the path used before and after the change? I suspect the problem is that the test now picks up a different "home dir" for which Flink does not have permission.

jiasheng55 commented:
@flinkbot run travis

jiasheng55 commented:
@kl0u this PR removed the "targetDir" parameter from `YarnClusterDescriptor#uploadAndRegisterFiles` method and use `FileSystem#getHomeDirectory` as the default target direcotry, so the `YarnFileStageTestS3ITCase .testRecursiveUploadForYarnS3a ()` failed becuase it seemed having credential error accessing the **"/user/{user.name}"** bucket.

If we choose not to use `mockito`, is it possible to set the bucket **"/user"**  accessible?

kl0u commented:
Hi @jiasheng55, can you rebase on the current master and push here? There seem to be more test failures https://api.travis-ci.org/v3/job/665755264/log.txt on the PR.

kl0u commented:
I am asking to rebase because the AZURE run was fixed and I think it also runs the S3 tests, so you can see if there is anything still failing.

kl0u commented:
The more I think about it, the more I lean towards leaving the `setupLocalResource()` as it is and add a method that simply calls this method with the `getHomeDirectory` as argument. In production we use the one without the `targetDir` as argument, and in the tests we use the old one. 

What do you think @jiasheng55 . This will allow us to get rid of the mocking as well.

jiasheng55 commented:
@flinkbot run travis

jiasheng55 commented:
@kl0u To get rid of the mocking, I rebased the current master branch and refactored this PR without touching the original flink-yarn tests.
 What do you think? Should I "add a `setupLocalResource` method without `targetDir` as parameter" as suggested?

tillrohrmann commented on flink-yarn/src/main/java/org/apache/flink/yarn/Utils.java:
```diff
@@ -99,6 +99,18 @@ public static void setupYarnClassPath(Configuration conf, Map<String, String> ap
+	 * @param appId
+	 * 		application ID
+	 */
+	public static Path getYarnFilesDir(final Path homedir, final String appId) {
```
I think we can tighten the contract a bit and say that `appId` must be of type `ApplicationId`.

tillrohrmann commented:
Has been subsumed by #11991.

kl0u commented:
Really sorry for that @jiasheng55 . It was my bad, I just got carried away while working on the other issue and did not inform you about my work. Sorry and I hope this will not discourage you in any way from contributing to Flink.

jiasheng55 commented:
@kl0u That's fine:) I will try to contribute more PRs to flink.
